// jest-cucumber/dist/src/feature-definition-creation.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var scenario_validation_1 = require("./validation/scenario-validation");
var step_definition_validation_1 = require("./validation/step-definition-validation");
var processScenarioTitleTemplate = function (scenarioTitle, parsedFeature, options, parsedScenario, parsedScenarioOutline) {
    if (options && options.scenarioNameTemplate) {
        try {
            return options && options.scenarioNameTemplate({
                featureTitle: parsedFeature.title,
                scenarioTitle: scenarioTitle.toString(),
                featureTags: parsedFeature.tags,
                scenarioTags: (parsedScenario || parsedScenarioOutline).tags
            });
        }
        catch (err) {
            throw new Error("An error occurred while executing a scenario name template. \nTemplate:\n" + options.scenarioNameTemplate + "\nError:" + err.message);
        }
    }
    return scenarioTitle;
};
var checkForPendingSteps = function (scenarioFromStepDefinitions) {
    var scenarioPending = false;
    scenarioFromStepDefinitions.steps.forEach(function (step) {
        try {
            if (step.stepFunction.toString().indexOf('pending()') !== -1) {
                var pendingTest = new Function("\n                    let isPending = false;\n\n                    const pending = function () {\n                        isPending = true;\n                    };\n\n                    (" + step.stepFunction + ")();\n\n                    return isPending;\n                ");
                scenarioPending = pendingTest();
            }
            else {
            }
        }
        catch (err) {
            //Ignore
        }
    });
    return scenarioPending;
};
var defineScenario = function (scenarioTitle, scenarioFromStepDefinitions, parsedScenario) {
    test(scenarioTitle, function () {
        return scenarioFromStepDefinitions.steps.reduce(function (promiseChain, nextStep, index) {
            var stepArgument = parsedScenario.steps[index].stepArgument;
            var matches = step_definition_validation_1.matchSteps(parsedScenario.steps[index].stepText, scenarioFromStepDefinitions.steps[index].stepMatcher);
            var matchArgs = [];
            if (matches && matches.length) {
                matchArgs = matches.slice(1);
            }
            var args = matchArgs.concat([stepArgument]);
            return promiseChain.then(function () { return nextStep.stepFunction.apply(nextStep, args); });
        }, Promise.resolve());
    });
};
var createDefineScenarioFunction = function (featureFromStepDefinitions, parsedFeature) {
    return function (scenarioTitle, stepsDefinitionFunctionCallback) {
        var scenarioFromStepDefinitions = {
            title: scenarioTitle,
            steps: []
        };
        featureFromStepDefinitions.scenarios.push(scenarioFromStepDefinitions);
        stepsDefinitionFunctionCallback({
            defineStep: createDefineStepFunction(scenarioFromStepDefinitions),
            given: createDefineStepFunction(scenarioFromStepDefinitions),
            when: createDefineStepFunction(scenarioFromStepDefinitions),
            then: createDefineStepFunction(scenarioFromStepDefinitions),
            pending: function () { }
        });
        var parsedScenario = parsedFeature.scenarios.filter(function (s) { return s.title === scenarioTitle; })[0];
        var parsedScenarioOutline = parsedFeature.scenarioOutlines.filter(function (s) { return s.title === scenarioTitle; })[0];
        var options = parsedFeature.options;
        scenarioTitle = processScenarioTitleTemplate(scenarioTitle, parsedFeature, options, parsedScenario, parsedScenarioOutline);
        step_definition_validation_1.ensureFeatureFileAndStepDefinitionScenarioHaveSameSteps(options, parsedScenario || parsedScenarioOutline, scenarioFromStepDefinitions);
        if (checkForPendingSteps(scenarioFromStepDefinitions)) {
            xtest(scenarioTitle, function () { }, undefined);
        }
        else if (parsedScenario) {
            defineScenario(scenarioTitle, scenarioFromStepDefinitions, parsedScenario);
        }
        else if (parsedScenarioOutline) {
            parsedScenarioOutline.scenarios.forEach(function (scenario) {
                defineScenario(scenarioTitle, scenarioFromStepDefinitions, scenario);
            });
        }
    };
};
var createDefineStepFunction = function (scenarioFromStepDefinitions) {
    return function (stepMatcher, stepFunction) {
        var stepDefinition = {
            stepMatcher: stepMatcher,
            stepFunction: stepFunction
        };
        scenarioFromStepDefinitions.steps.push(stepDefinition);
    };
};
function defineFeature(featureFromFile, scenariosDefinitionCallback) {
    var featureFromDefinedSteps = {
        title: featureFromFile.title,
        scenarios: []
    };
    scenariosDefinitionCallback(createDefineScenarioFunction(featureFromDefinedSteps, featureFromFile));
    scenario_validation_1.checkThatFeatureFileAndStepDefinitionsHaveSameScenarios(featureFromFile, featureFromDefinedSteps);
}
exports.defineFeature = defineFeature;
